### ISSUE TITLE
[iOS] Pixelated svg icons

### ISSUE THREAD

**Issue by Primo85** [dnfield/flutter_svg#975]
Since we are using 'color' parameter there is a problem with svg rendering on iOS (on android it works fine).
The problem is in tab bar only if there are 2 items there. For 3 and more elements it works fine.
Tested on iPhone 7 iOS 15.7,  iPhone 8 iOS 16.3.0 and iPhone 8 iOS 16.3.1.

<details><summary>Expected results/Screenshot</summary>

![OK](https://github.com/dnfield/flutter_svg/assets/23534114/10f16af9-71aa-4114-b5d4-0c211ae43804)

</details>
<details><summary>Actual results/Screenshot</summary>

![NOK2](https://github.com/dnfield/flutter_svg/assets/23534114/e4ec94d5-8b0d-4235-876e-0435558ad2ef)
![OK3](https://github.com/dnfield/flutter_svg/assets/23534114/ac7af944-b3a5-4ff2-b805-aed506838a36)

</details>
<details><summary>main.dart</summary>

```dart
import 'package:flutter/material.dart';
import 'package:flutter_svg/svg.dart';

void main() {
  runApp(const MyApp());
}
class MyApp extends StatelessWidget {
  const MyApp({super.key});

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Demo',
      theme: ThemeData(primarySwatch: Colors.blue),
      home: const MyHomePage(),
    );
  }
}

class MyHomePage extends StatefulWidget {
  const MyHomePage();

  @override
  State<MyHomePage> createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  int index = 0;

  void _setIndex(int index) {
    setState(() {
      this.index = index;
    });
  }

  List<BottomNavigationBarItem> getTabs(BuildContext context) {
    return [
      BottomNavigationBarItem(
        icon: Assets.home,
        label: "item 1",
      ),
      BottomNavigationBarItem(
        icon: Assets.homeColor,
        label: "item 2",
      ),
      // BottomNavigationBarItem(
      //   icon: Assets.homeColor,
      //   label: "item 3",
      // ),
    ];
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(title: Text("Some title")),
      body: const Center(
        child: Text("Some screen"),
      ),
      bottomNavigationBar: BottomNavigationBar(
        type: BottomNavigationBarType.fixed,
        onTap: (int index) {
          _setIndex(index);
        },
        currentIndex: index,
        items: getTabs(context),
      ),
    );
  }
}

class Assets {
  static Color filterColor = Color.fromRGBO(0, 0, 0, 1);

  static Widget home = SvgPicture.asset("assets/home.svg");

  static Widget homeColor = SvgPicture.asset("assets/home.svg",
      color: filterColor);
}
```

</details>
<details><summary>home.svg</summary>

![home](https://github.com/dnfield/flutter_svg/assets/23534114/ba2713bb-7b71-4edd-9b1f-c268c96316a4)

</details>
<details><summary>Doctor output</summary>

```console
Doctor summary (to see all details, run flutter doctor -v):
[!] Flutter (Channel master, 3.11.0-14.0.pre.6, on macOS 13.1 22C65 darwin-x64, locale en-PL)
    ! Warning: `dart` on your path resolves to /usr/local/Cellar/dart/2.18.7/libexec/bin/dart, which is not inside your current Flutter SDK checkout at /Users/pbiskup/fvm/versions/3.10.1. Consider adding /Users/pbiskup/fvm/versions/3.10.1/bin to the front of your path.
[!] Android toolchain - develop for Android devices (Android SDK version 33.0.2)
    ✗ cmdline-tools component is missing
      Run `path/to/sdkmanager --install "cmdline-tools;latest"`
      See https://developer.android.com/studio/command-line for more details.
    ✗ Android license status unknown.
      Run `flutter doctor --android-licenses` to accept the SDK licenses.
      See https://flutter.dev/docs/get-started/install/macos#android-setup for more details.
[✓] Xcode - develop for iOS and macOS (Xcode 14.3)
[✓] Chrome - develop for the web
[✓] Android Studio (version 2022.1)
[✓] VS Code (version 1.79.2)
[✓] Connected device (3 available)
[✓] Network resources
```

</details>

**Comment by maximletushov** [comment]
Hi,
I was faces with the same issue with `flutter_svg: 2.0.7` and would like to add some additional info:
1) results on real iOS device much worse than on iOS Simulator
2) results depends on the size of the output image
My svg has parameters `width="24"` `height="24"` and `viewBox="0 0 24 24"` and I received the worst results for sizes 20, 60, 100.

<details><summary>Results on real iOS Device</summary>
<p>

![Device - iPhone 14 Pro Max](https://github.com/dnfield/flutter_svg/assets/839819/472d9fbb-1c27-4508-a14e-554b325f2f7c)

</p>
</details>

<details><summary>Results on iOS Simulator</summary>
<p>

![Simulator - iPhone 14 Pro Max](https://github.com/dnfield/flutter_svg/assets/839819/5a3946c7-5454-4276-8200-1216f370aac9)

</p>
</details> 

<details><summary>Svg file</summary>
<p>

![alert-circle](https://github.com/dnfield/flutter_svg/assets/839819/23d4e87c-20e4-4e9e-9b3f-09b38c99a792)

</p>
</details> 

<details><summary>Demo code</summary>
<p>

```dart
Widget _buildContentForDemo(BuildContext context, String assetName, String package) {
    const List<double> sizes = [16, 20, 24, 30, 40, 50, 60, 80, 100, 120];
    const colorFilter = ColorFilter.mode(Colors.green, BlendMode.srcIn);

    return SafeArea(
      child: SingleChildScrollView(
        child: Padding(
          padding: const EdgeInsets.all(24),
          child: Column(
            children: sizes.map((size) {
              return Padding(
                padding: const EdgeInsets.only(bottom: 24),
                child: Row(
                  children: [
                    SizedBox(width: 50, child: Text('$size')),
                    SvgPicture.asset(
                      assetName,
                      package: package,
                      colorFilter: null, // without color
                      width: size,
                      height: size,
                    ),
                    const SizedBox(width: 24),
                    SvgPicture.asset(
                      assetName,
                      package: package,
                      colorFilter: colorFilter, // with color
                      width: size,
                      height: size,
                    ),
                  ],
                ),
              );
            }).toList(),
          ),
        ),
      ),
    );
  }
``` 

</p>
</details> 

<details><summary>Flutter doctor</summary>
<p>

```
[✓] Flutter (Channel stable, 3.10.6, on macOS 13.3.1 22E772610a darwin-x64, locale en-UA)
[✓] Android toolchain - develop for Android devices (Android SDK version 33.0.1)
[✓] Xcode - develop for iOS and macOS (Xcode 14.3.1)
[✓] Chrome - develop for the web
[✓] Android Studio (version 2022.3)
```
</p>
</details> 

Also I think this issue is very similar to #927 

Please take a look

**Comment by neronovs** [comment]
Same for me. I use the latest 2.0.7 version. 
Android is rendering as expected:
<img width="1712" alt="android Screenshot 2023-08-15 at 11 05 31" src="https://github.com/dnfield/flutter_svg/assets/19669844/7bf73f2e-f54f-4de2-b67f-eeee04cb5586">
iOS lags:
<img width="1679" alt="ios Screenshot 2023-08-15 at 11 04 52" src="https://github.com/dnfield/flutter_svg/assets/19669844/6063565e-33ad-48d4-bdbd-55372fb28435"> 

It would be super to fix this kind of issue!

**Comment by maximletushov** [comment]
@dnfield thank you for response

I've checked on beta and even on master channel and the problem still exists.

<details><summary>result on beta channel</summary>
<p>

![beta channel](https://github.com/dnfield/flutter_svg/assets/839819/9952c6e0-7318-49ab-9446-8dc97d749ae9)

</p>
</details> 


<details><summary>doctor on beta</summary>
<p>

```
[✓] Flutter (Channel beta, 3.13.0-0.4.pre, on macOS 13.3.1 22E772610a darwin-x64, locale en-UA)
[✓] Android toolchain - develop for Android devices (Android SDK version 33.0.1)
[✓] Xcode - develop for iOS and macOS (Xcode 14.3.1)
[✓] Chrome - develop for the web
[✓] Android Studio (version 2022.3)

```
</p>
</details> 


<details><summary>result on master channel</summary>
<p>

![master channel](https://github.com/dnfield/flutter_svg/assets/839819/9d4f67fa-786c-40bd-a43f-ce8f1140a66a)


</p>
</details> 

<details><summary>doctor on master</summary>
<p>

```
[✓] Flutter (Channel master, 3.14.0-5.0.pre.27, on macOS 13.3.1 22E772610a darwin-x64, locale en-UA)
[✓] Android toolchain - develop for Android devices (Android SDK version 33.0.1)
[✓] Xcode - develop for iOS and macOS (Xcode 14.3.1)
[✓] Chrome - develop for the web
[✓] Android Studio (version 2022.3)

```
</p>
</details>

**Comment by Primo85** [comment]
I have checked it too, on the master, the problem still occurs.

**Comment by aytunch** [comment]
Hi,
The moment I remove the `colorFilter: ColorFilter.mode(color, BlendMode.srcIn)` field, this problem goes away.
But of course, when we do this, the SVGs fallback to their natural color rendering the app useless.
The svgs have such jagged lines that the whole app looks really bad.
Is this issue caused by Impeller or this plugin? Should we open up an issue in Flutter/engine repo?

**Comment by cliftonlabrum** [comment]
Any progress with this one?

**Comment by dnfield** [comment]
I don't have a good reproduction case for this. I need one htat I can reproduce locally to be able to work on this.

**Comment by cliftonlabrum** [comment]
Here is a reproduction of the issue. I am using Xcode 15 with an iPhone 15 Pro simulator on `Flutter 3.13.5`.

### pubspec.yaml
```yaml
name: flutter_svg_rendering_bug
description: A new Flutter project.
publish_to: 'none'
version: 1.0.0+1
environment:
  sdk: '>=3.1.2 <4.0.0'
dependencies:
  flutter:
    sdk: flutter
  cupertino_icons: ^1.0.2
  flutter_svg: ^2.0.7
dev_dependencies:
  flutter_test:
    sdk: flutter
  flutter_lints: ^2.0.0
flutter:
  assets:
    - assets/menu.svg
    - assets/add.svg
```
### main.dart
```dart
import 'package:flutter/material.dart';
import 'package:flutter_svg/flutter_svg.dart';

void main() {
  runApp(const MyApp());
}

class MyApp extends StatelessWidget {
  const MyApp({super.key});

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      home: Center(
        child: SizedBox(
          width: 200,
          height: 200,
          child: Column(
            children: [
              Row(
                children: [
                  SvgPicture.asset(
                    'assets/menu.svg',
                    colorFilter: const ColorFilter.mode(Colors.orange, BlendMode.srcIn),
                    width: 20,
                    height: 16,
                  ),
                  const SizedBox(width: 30),
                  SvgPicture.asset(
                    'assets/add.svg',
                    colorFilter: const ColorFilter.mode(Colors.orange, BlendMode.srcIn),
                    width: 24,
                    height: 24,
                  ),
                ],
              ),
              const SizedBox(height: 30),
              Row(
                children: [
                  SvgPicture.asset(
                    'assets/menu.svg',
                    //colorFilter: const ColorFilter.mode(Colors.red, BlendMode.srcIn),
                    width: 20,
                    height: 16,
                  ),
                  const SizedBox(width: 30),
                  SvgPicture.asset(
                    'assets/add.svg',
                    //colorFilter: const ColorFilter.mode(Colors.orange, BlendMode.srcIn),
                    width: 24,
                    height: 24,
                  ),
                ],
              ),
            ],
          ),
        ),
      ),
    );
  }
}
```
Attached are the two SVG assets to include.

![add](https://github.com/dnfield/flutter_svg/assets/2513753/f45a3bb0-d744-4621-ab8a-82aaeb5e52b9)
![menu](https://github.com/dnfield/flutter_svg/assets/2513753/ffb6f1fe-f7e5-4f39-ab6e-67766a10bf1b)

The screenshots attached show that the orange icons have a `colorFilter` property set, and the blue ones do not (the blue color is from the SVG itself).

<img width="140" alt="1" src="https://github.com/dnfield/flutter_svg/assets/2513753/d8607802-f836-4052-abeb-683a44090156">

<img width="122" alt="2" src="https://github.com/dnfield/flutter_svg/assets/2513753/c9fcb6ed-415b-4578-991c-ce7538a03caf">

![3](https://github.com/dnfield/flutter_svg/assets/2513753/11a22f2a-9342-4fd8-a772-d0969b0583c7)


Hopefully you can see how the edges of the orange icons is jagged and distorted. Let me know you need anything else! Thanks for your help!

**Comment by Primo85** [comment]
Hi. 
Another example of reproduction is in my first post, in the main.dart tab. I'm using Xcode 14.3 and Flutter 3.10.6(and also Channel master, 3.11.0-14.0.pre.6).
Thank you.

**Comment by armandsLa** [comment]
Having the same issue. Interestingly, only with a single icon out of ~50. Was fine on iOS 16. 

Here is the icon:
![ic_nav_close](https://github.com/dnfield/flutter_svg/assets/6855713/98866777-02fd-497f-b6fe-8ca61acf1809)

**Comment by simonNEON** [comment]
Are there any updates on this? Im having the same problems and not yet found a solution...

**Comment by SivaramSS** [comment]
Facing the same issue with flutter 3.13.9.

**Comment by Pomis** [comment]
Flutter 3.13.8 with Impeller, all SVGs look jagged

**Comment by armandsLa** [comment]
I've found a temporary fix. Wrapping `SvgPicture` in `Opacity` and setting the `opacity` to `0.99` fixes the issue. Opacity reduction of 1% is visually unnoticeable.

```
Opacity(
  opacity: 0.99,
  child: SvgPicture.asset("asset/img_check.svg",
    colorFilter: ColorFilter.mode(
      Colors.red,
      BlendMode.srcIn,
    ),
  ),
);
```

**Comment by LeonardoRosaa** [comment]
> I've found a temporary fix. Wrapping `SvgPicture` in `Opacity` and setting the `opacity` to `0.99` fixes the issue. Opacity reduction of 1% is visually unnoticeable.
> 
> ```
> Opacity(
>   opacity: 0.99,
>   child: SvgPicture.asset("asset/img_check.svg",
>     colorFilter: ColorFilter.mode(
>       Colors.red,
>       BlendMode.srcIn,
>     ),
>   ),
> );
> ```

Not sure about this solution. After some hours of debugging this issue, I noticed that the issue could be in the `colorFilter` parameter (the new way to set colors).

Wrapping the icon with the `Opacity` widget made all icons more "pixelated" in my case.

**Comment by RicharC293** [comment]
Same issue here

**Comment by dimazelinskyi** [comment]
<img width="330" alt="Screenshot 2023-12-10 at 12 57 38" src="https://github.com/dnfield/flutter_svg/assets/1001657/0fc85a15-54e7-4022-ae9f-12d6d139fe8c">

It looks like related issue  https://github.com/psk907/fluttermoji/issues/35

**Comment by HardikKotadiya** [comment]
Facing the same issue with flutter 3.16.5. I had tried lots of solution but not worked, please let me know if anyone get solution.

I had already post issue in stackoverflow too at below link

https://stackoverflow.com/questions/77741764/flutter-svg-icons-get-pixelated-in-bottom-bar-in-flutter

**Comment by kienvtqhi** [comment]
I got the same issue

**Comment by haomun0731** [comment]
This issue occurs on my iOS simulator but not on the real device. I temporarily fixed it by disabling impeller rendering in iOS debug mode. I added the following lines to `Info-Debug.plist`:

```
<key>FLTEnableImpeller</key>
<false />
```

For more information, please refer to: https://docs.flutter.dev/perf/impeller#ios

**Comment by capsulekenny** [comment]
Any updates on this?

**Comment by armandsLa** [comment]
This was clearly an issue with Flutter itself. Just run the test with `3.19.0` version and icons look good.

**Comment by vase4kin** [comment]
I can confirm. It is gone. The flutter version I tested with is `3.19.1`.

**Comment by felipecastrosales** [comment]
I can confirm, in latest Flutter versions (3.19.x) this issue isn't present. I think also that this issue can be closed.

> This was clearly an issue with Flutter itself. Just run the test with `3.19.0` version and icons look good.

**Comment by anatoolybinerals** [comment]
And how fix this issue in 2024?

flutter_svg: ^1.1.6

In my case help not use `color` property:
```
SvgPicture.asset(
  SvgAssetName.chat,
  // color: AppColors.bottomBarAppIcon,
)
```

**Comment by LeonardoRosaa** [comment]
The issue is back in the Flutter version `3.19.4`. Some icons are pixelated in this Flutter version.

**Comment by Kal-Elx** [comment]
I can also confirm that the issue is back in later Flutter versions. 
@dnfield Do you know if this is currently being worked on?

**Comment by armandsLa** [comment]
Fortunately, the patch is still working. 

```
import 'dart:io';
import 'package:flutter/material.dart';
import 'package:flutter_svg/svg.dart';

class SvgIcon extends StatelessWidget {

  final String path;
  final Color color;

  final double? width;
  final double? height;

  SvgIcon({
    required this.path,
    required this.color,
    this.width,
    this.height,
  });

  @override
  Widget build(BuildContext context) {
    if (Platform.isAndroid) {
      return _buildImage();
    }
    return Transform.scale(
      scale: 0.99,
      child: _buildImage(),
    );
  }

  Widget _buildImage() {
    final colorFilter = ColorFilter.mode(
      color,
      BlendMode.srcIn,
    );
    return SvgPicture.asset(path,
      colorFilter: colorFilter,
      width: width,
      height: height,
    );
  }

}
```

**Comment by LiuMentz** [comment]
Still present with Flutter 3.22.0

**Comment by EgorK0rshun** [comment]
Still present with Flutter 3.22.1

**Comment by felipecastrosales** [comment]
Hey guys, I see that @dnfield has been inactive on GitHub for a while now. So, I don't think he's seen the latest on this issue. 

Another person I see who collaborates on the package and is very close to the Flutter team is @jonahwilliams, who could possibly give us some guidance on what to do here - or check what might have caused the error to happen again.

@Primo85, please reopen the issue so that this can be given some priority.

**Comment by johnthiriet** [comment]
Hello, I can also confirm that the issue is present in Flutter 3.22.1
It's not reproduced on the main channel though.
The scale hack still works

**Comment by maxAtIppen** [comment]
I also have that issue with flutter_svg 2.0.10+1 and flutter 3.22.2

**Comment by bahadurh** [comment]
> > Fortunately, the patch is still working.
> 
> > 
> 
> > ```
> 
> > import 'dart:io';
> 
> > import 'package:flutter/material.dart';
> 
> > import 'package:flutter_svg/svg.dart';
> 
> > 
> 
> > class SvgIcon extends StatelessWidget {
> 
> > 
> 
> >   final String path;
> 
> >   final Color color;
> 
> > 
> 
> >   final double? width;
> 
> >   final double? height;
> 
> > 
> 
> >   SvgIcon({
> 
> >     required this.path,
> 
> >     required this.color,
> 
> >     this.width,
> 
> >     this.height,
> 
> >   });
> 
> > 
> 
> >   @override
> 
> >   Widget build(BuildContext context) {
> 
> >     if (Platform.isAndroid) {
> 
> >       return _buildImage();
> 
> >     }
> 
> >     return Transform.scale(
> 
> >       scale: 0.99,
> 
> >       child: _buildImage(),
> 
> >     );
> 
> >   }
> 
> > 
> 
> >   Widget _buildImage() {
> 
> >     final colorFilter = ColorFilter.mode(
> 
> >       color,
> 
> >       BlendMode.srcIn,
> 
> >     );
> 
> >     return SvgPicture.asset(path,
> 
> >       colorFilter: colorFilter,
> 
> >       width: width,
> 
> >       height: height,
> 
> >     );
> 
> >   }
> 
> > 
> 
> > }
> 
> > ```
> 
> 
> 
> This workaround works for me in debug mode but not in release mode. Does it work for you all in release mode?

Play with transform, for me 1.02 scale worked fine 👍🏻👍🏻

**Comment by maxAtIppen** [comment]
I tried 0.99, 1.01 and 1.02 but all of them still look pixelated.

**Comment by Joan1211** [comment]
Same issue with flutter 3.22.2
Issue only on iOS.

Even values for the size (height:24, width: 24) : issue is not visible
Odd values for the size (height:23, width: 23) : issue is visible

**Comment by SimoneScaglia** [comment]
Same issue.

I solved it removing colorFilter parameter and specifying the color directly in svg code.

**Comment by emirhankolver** [comment]
I encountered an issue where icons appeared pixelated when using the `ColorFilter` parameter with the `SvgPicture.asset` widget. Initially, I tried using `Transform.scale`, but unfortunately, the icons still looked pixelated. Then, I added the `filterQuality: FilterQuality.low` parameter, which resolved the issue and the icons started to look great again.

Here’s my codebase:

```dart
class IconView extends StatelessWidget {
  final IconAssetData? asset;
  final double? size;
  final Color? color;

  const IconView({
    super.key,
    this.asset,
    this.color,
    this.size = 24,
  });

  @override
  Widget build(BuildContext context) {
    if (asset == null) {
      return SizedBox(width: size, height: size);
    }
    final iconColor = color ?? context.theme.colorScheme.onSurface;
    if (Platform.isAndroid) {
      return _renderSVG(iconColor);
    }
    return Transform.scale(
      filterQuality: FilterQuality.low,
      scale: 0.9999,
      child: _renderSVG(iconColor),
    );
  }

  Widget _renderSVG(Color iconColor) {
    return SvgPicture.asset(
      asset!.path,
      width: size,
      height: size,
      colorFilter: ColorFilter.mode(iconColor, BlendMode.srcIn),
    );
  }
}
```


Previews:
1. **filterQuality: FilterQuality.none**
![Ekran Resmi 2024-07-28 - 13 18 42](https://github.com/user-attachments/assets/7b00d9d7-87d3-4ac4-a03b-74d3e2b74edb)

2. **filterQuality: FilterQuality.high**
![IMG_DBE3D0063EA2-1](https://github.com/user-attachments/assets/827f05dc-fb3b-41c4-9313-e6a7c0c14c89)

**Comment by qwertylolman** [comment]
it's still present in the package :(
I have to replace colorFilter with colorMapper

**Comment by beingsalmanshaikh** [comment]
The issue is still present. Even the workaround is not working for some reason.

**Comment by StolarczykJakub** [comment]
Flutter (Channel stable, 3.22.2)
flutter_svg: ^2.0.10+1
iPhone 16 pro (on mini 12 its good, seems like resolution is problem)

seems to occure again, without any color tint, just `SvgPicture.asset(profileIcon),`

**Comment by chriscarman-mp** [comment]
I can't get any of the fixes to work - Flutter 3.13.9 and flutter_svg 2.0.7

Driving me nuts...

**Comment by mentz-lwa** [comment]
@chriscarman-mp try adding a padding inside the svg.
